src/mission_companion.cpp

```diff
@@ -1420,23 +3400,128 @@ void talk_function::companion_return( npc &comp ){
+    comp.companion_mission_time_ret = calendar::before_time_starts;
+    for( size_t i = 0; i < comp.companion_mission_inv.size(); i++ ) {
+        for( const auto &it : comp.companion_mission_inv.const_stack( i ) ) {
+            g->m.add_item_or_charges( g->u.posx(), g->u.posy(), it );
+        }
+    }
+    comp.companion_mission_inv.clear();
     // npc *may* be active, or not if outside the reality bubble
     g->reload_npcs();
 }
 
-std::vector<std::shared_ptr<npc>> talk_function::companion_list( const npc &p, const std::string &id )
+std::vector<std::shared_ptr<npc>> talk_function::companion_list( const npc &p, const std::string &id, bool contains )
 {
     std::vector<std::shared_ptr<npc>> available;
+    const point omt_pos = ms_to_omt_copy( g->m.getabs( p.posx(), p.posy() ) );
+    std::string mission_format = string_format( "%s(%d,%d,%d)%s", p.companion_mission_role_id, omt_pos.x, omt_pos.y, p.posz(), id );
     for( const auto &elem : overmap_buffer.get_companion_mission_npcs() ) {
-        if( elem->get_companion_mission() == p.name + id ) {
+        if( elem->get_companion_mission() == p.name + id || elem->get_companion_mission() == mission_format ) {
+            available.push_back( elem );
+        } else if( contains && (elem->get_companion_mission().find(p.name + id) != std::string::npos ||
+                  (elem->get_companion_mission().find(mission_format) != std::string::npos ) ) ){
             available.push_back( elem );
         }
     }
     return available;
 }
 
-npc *talk_function::companion_choose(){
-    const std::vector<npc *> available = g->get_npcs_if( [&]( const npc &guy ) {
+bool companion_sort_compare( npc* first, npc* second ){
+    std::vector<npc *> comp_list;
+    comp_list.push_back(first);
+    comp_list.push_back(second);
+    std::vector<tripoint> scores = talk_function::companion_rank( comp_list );
+    if( scores[0].x > scores[1].x ){
+        return true;
+    }
+    return false;
+}
+
+std::vector<npc *> talk_function::companion_sort( std::vector<npc *> available, std::string skill_tested )
+{
+    if( skill_tested.empty() ){
+        std::sort (available.begin(), available.end(), companion_sort_compare);
+    } else {
+
+        struct companion_sort_skill {
+            companion_sort_skill(std::string skill_tested) { this->skill_tested = skill_tested; }
+            //bool operator () (int i, int j) { ... }
+
+            bool operator()( npc* first, npc* second ){
+                if( first->get_skill_level( skill_id(skill_tested) ) > second->get_skill_level( skill_id(skill_tested) ) ){
+                        return true;
+                }
+                return false;
+            }
+
+
+            std::string skill_tested;
+        };
+        std::sort (available.begin(), available.end(), companion_sort_skill(skill_tested) );
+    }
+    return available;
+}
+
+std::vector<tripoint> talk_function::companion_rank( std::vector<npc *> available, bool adj )
```

Comment:
`tripoint` should only ever describe position.
Create a new `struct` so that you know what do the values mean.